cli/backend_base.py

```diff
@@ -207,14 +208,65 @@ def _ship_certs(self, cluster, saltmaster_ip):
 
         return platform_certs_tarball
 
+    def _fill_volume_config(self, volume_dict):
+        ''' returns the generated volumes for kafka and datanodes '''
+        data_volume_list = []
+        if self._pnda_env['disk_volumes']['GENERATE_VOLUME_CONFIG'] == "YES":
+            datanode_volume_count = int(self._pnda_env['disk_volumes']['DATANODE_VOLUME_COUNT'])
+            if datanode_volume_count == 0:
+                CONSOLE.error('datanode volume count should not be zero')
+                sys.exit(1)
+
+            for volume in volume_dict['volumes']:
+                if '/data' in volume:
+                    volume_index = volume_dict['volumes'].index(volume)
+                    del volume_dict['volumes'][volume_index]
+
+            if datanode_volume_count > 10:
+                prefix = "/data0"
+            else:
+                prefix = "/data"
+
+            for dir_count  in range(datanode_volume_count):
+                data_dir = "%s%s" %(prefix, dir_count)
+                data_dir_fs = "%s xfs" %data_dir
+                volume_dict['volumes'].append(data_dir_fs)
+                data_volume_list.append(data_dir+"/dn")
+        else:
+            for volume in volume_dict['volumes']: 
+                if '/data' in volume:
+                    if len(volume.split(" ")) == 2:
+                        data_volume_list.append(volume.split(" ")[0]+"/dn")
+                    if len(volume.split(" ")) == 3:
+                        data_volume_list.append(volume.split(" ")[1]+"/dn")
+            if not data_volume_list:
+                CONSOLE.error('data volume count should not be zero')
+                sys.exit(1)
+
+        return volume_dict, data_volume_list
+
     def _get_volume_info(self, node_type, config_file):
+        ''' returns volume info for all instances '''
         volumes = None
         if node_type:
             with open(config_file, 'r') as infile:
                 volume_config = yaml.load(infile)
                 volume_class = volume_config['instances'][node_type]
                 volumes = volume_config['classes'][volume_class]
-        return volumes
+
+        if 'hadoop-dn' in node_type:
+            dn_volume_dict = copy.deepcopy(volumes)
+            ret_volumes, data_volumes_list = self._fill_volume_config(dn_volume_dict)
+            return ret_volumes, data_volumes_list
+
+        if node_type == self._node_config['salt-master-instance']:
```

Comment:
I can't follow the purpose of this code, why does the salt master need to use the volume class for a hadoop-dn. Could you add some comments to explain in simple terms what's going on here?

Reply:
Hi James,
Have addressed all the comments.
Pylint score is 10/10.
Salt-master uses hadoop-dn class to export the generated datanode disk directories, to use in platform salt as a pillar.